### PR TITLE
Make docs tasks cacheable

### PR DESCRIPTION
**Pull request by jjohannes**
Make the doc tasks cacheable to allow building the `all` distribution from cache.

### CODE REVIEW

**Comment by jjohannes** [comment]
I submitted a PR to the Asciidoc Gradle Plugin to make it cacheable without the hack: https://github.com/asciidoctor/asciidoctor-gradle-plugin/pull/215

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy [hunk 1]
```diff
@@ -51,9 +56,14 @@ class UserGuideSectionVerifier extends DefaultTask {
             appendMissingIdsMessagePart(exceptionMessage, allResults)
         }
 
+        reportFile.parentFile.mkdirs()
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy]: I get `No such file or directory` without this.

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy [hunk 2]
```diff
@@ -5,29 +5,34 @@ import groovy.transform.ToString
...
  * duplicate section IDs.
  */
 @CompileStatic
+@CacheableTask
 class UserGuideSectionVerifier extends DefaultTask {
     @InputFiles
     FileCollection docbookFiles
 
     @Input
     String fileEncoding = "UTF-8"
```
Comment (wolfs) [inline buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy]: File encoding is shown as "unused" in IDEA. Where is that required?

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjohannes) [reply buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy]: `UserGuideSectionVerifierTest` told me I need this :)

Thanks for asking. Removed.

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy [hunk 3]
```diff
@@ -51,9 +56,14 @@ class UserGuideSectionVerifier extends DefaultTask {
             appendMissingIdsMessagePart(exceptionMessage, allResults)
         }
 
+        reportFile.parentFile.mkdirs()
+        reportFile.createNewFile()
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy]: Removed.

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy (hunk 2 above)
Comment (jjohannes) [reply buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy]: Right. It's not used. Removed it.

File: buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy [hunk 4]
```diff
@@ -5,29 +5,34 @@ import groovy.transform.ToString
...
-
 /**
  * This is a Gradle task that runs through the user guide's source XML files
  * and checks that all sections have {@code id} attributes and there are no
  * duplicate section IDs.
  */
 @CompileStatic
+@CacheableTask
 class UserGuideSectionVerifier extends DefaultTask {
     @InputFiles
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply buildSrc/src/main/groovy/org/gradle/build/docs/UserGuideSectionVerifier.groovy]: I added PathSensitivity.NAME_ONLY

File: subprojects/docs/docs.gradle [hunk 5]
```diff
@@ -123,7 +124,21 @@ outputs.docs = files(docsDir) {
     builtBy 'javadocAll', 'userguide', 'dslHtml', 'releaseNotes'
 }
 
-tasks.withType(AsciidoctorTask) {
+@CacheableTask
+class CacheableAsciidoctorTask extends AsciidoctorTask {
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply subprojects/docs/docs.gradle]: Yes, moving this to `buildSrc` is better. I did that.

All the inputs are marked as `@Optional`. I added `PathSensitivity.RELATIVE` to the once we are using. Would that be sufficient?

**Comment by jjohannes** [comment]
Thanks for the feedback @wolfs! Can you have another look. At particular at the PathSensitivity stuff?

File: subprojects/docs/docs.gradle [hunk 6]
```diff
@@ -123,7 +120,7 @@ outputs.docs = files(docsDir) {
     builtBy 'javadocAll', 'userguide', 'dslHtml', 'releaseNotes'
 }
 
-tasks.withType(AsciidoctorTask) {
+tasks.withType(CacheableAsciidoctorTask) {
```
Comment (wolfs) [inline subprojects/docs/docs.gradle]: We also need to add a path sensitivity to this file line: 
```
inputs.file asciidocSanitizeStylesheet withPropertyName "sanitizerStylesheet"
```
I guess `NONE`.

File: subprojects/docs/docs.gradle [hunk 7]
```diff
@@ -42,6 +38,7 @@ apply plugin: 'base'
 apply plugin: 'pegdown'
 apply plugin: 'jsoup'
 apply plugin: 'javascript-base'
+apply plugin: 'org.asciidoctor.convert'
```
Comment (wolfs) [inline subprojects/docs/docs.gradle]: Do we still need to apply the plugin or do we only need the types? We should get those from the buildSrc dependencies.

File: buildSrc/src/main/groovy/org/gradle/build/docs/CacheableAsciidoctorTask.groovy [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import org.asciidoctor.gradle.AsciidoctorTask
+import org.gradle.api.file.FileTree
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+
+@CompileStatic
+@CacheableTask
+class CacheableAsciidoctorTask extends AsciidoctorTask {
```
Comment (wolfs) [inline buildSrc/src/main/groovy/org/gradle/build/docs/CacheableAsciidoctorTask.groovy]: What happened to `StreamClosingAsciidoctorTask`?

Comment (jjohannes) [reply buildSrc/src/main/groovy/org/gradle/build/docs/CacheableAsciidoctorTask.groovy]: This PR is based on release. The `StreamClosingAsciidoctorTask` is on master. I would merge the stream closing part into `CacheableAsciidoctorTask` when I am merging this PR.

It should only be a temporary solution anyway. I am planing to look into that some more to see if I can find the real issue and solve it in a more elegant way. Or produce a upstream fix.

File: subprojects/docs/docs.gradle (hunk 7 above)
Comment (jjohannes) [reply subprojects/docs/docs.gradle]: Yes it does some setup. Creating the extension, a configuration and some dependency management. I would leave the `apply` in to use as much of the original plugin as possible.

File: subprojects/docs/docs.gradle (hunk 6 above)
Comment (jjohannes) [reply subprojects/docs/docs.gradle]: Done.

Also tested it now with two repository clones in different locations.

Comment (wolfs) [reply subprojects/docs/docs.gradle]: For a full test, you would also need to use a different Gradle user home directory to have the dependencies in some other place.

**Pull request merged**